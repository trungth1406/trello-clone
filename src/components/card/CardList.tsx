import Card from "./Card";
import {useState} from "react";
import {CardModel} from "../../models/card-model";

const DEFAULT_CARD: CardModel = {
    name: null,
    items: []
}

export default function CardList() {
    const [cards, setCards] = useState<CardModel[]>([DEFAULT_CARD]);
    return (
        <ul className='card-list'>
            {cards.map((card, index) => <Card key={index} card={card}/>)}
        </ul>)
}
