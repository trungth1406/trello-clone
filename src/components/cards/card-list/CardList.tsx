import Card from "../Card";
import {useState} from "react";
import {CardModel} from "../../../models/card-model";
import AddCardButton from "../AddCardButton";

const DEFAULT_CARD: CardModel = {
    id: Math.random(),
    name: 'Default Card',
    items: []
}

export default function CardList() {
    const [cards, setCards] = useState<CardModel[]>([DEFAULT_CARD]);

    const addCard = () => {
        setCards(last => [...last, {
            id: Math.random(),
            name: null,
            items: []
        }]);
    }

    const deleteCard = (card: CardModel) => {
        setCards(lastCards => [...lastCards.filter(c => c !== card)]);
    }

    return (
        <ul className='card-list'>
            {cards.map((card, index) => <Card  key={'cardNo' + card.id} card={card} onCardDelete={deleteCard}/>)}
            <AddCardButton  onAddNewCard={addCard}/>
        </ul>)
}
