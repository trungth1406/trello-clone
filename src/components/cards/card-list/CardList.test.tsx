// test that it Should render a single list of card\
//
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import CardList from './CardList';

describe('When CardList mounted', () => {
    const cardListComponent = renderer.create(<CardList />);
    const cardListTree = cardListComponent.toJSON();

    it('Should render a circle button to add a card', async () => {
        render(<CardList />);
        expect(cardListTree['children']).toBeTruthy();
        expect(cardListTree['type']).toBe('ul');
        expect(cardListTree['children']).toHaveLength(2);

        // expect to find a button with id 'add-card_btn'
        expect(screen.getByTestId('add-card_btn')).toBeTruthy();
        expect(screen.getByTestId('add-card_btn')).toBeInTheDocument();
    });

    it('Should render a default card', async () => {
        render(<CardList />);
        expect(screen.getAllByTestId('card')).toHaveLength(1);
        // expect the card to have the header  equal to 'Default Card' with screen
        expect(screen.getByText('Default Card')).toBeInTheDocument();

        expect(screen.getByTestId('delete_btn')).toBeTruthy();
        expect(screen.getAllByTestId('delete_btn')).toHaveLength(1);
        expect(cardListTree).toMatchSnapshot();
    });

    // When click on the add card button, it should add a new card
    it('Should add a new card when click on the add card button', async () => {
        // click on the add card button

        render(<CardList />);
        fireEvent.click(screen.getByTestId('add-card_btn'));
        expect(screen.getAllByTestId('card')).toHaveLength(2);
    });

    // When click on the delete button, it should delete the card
    it('Should delete the card when click on the delete button', async () => {
        render(<CardList />);
        fireEvent.click(screen.getByTestId('delete_btn'));
        expect(screen.queryAllByTestId('card')).toHaveLength(0);
    });

    describe('when done giving task a name and click on the add task  button', () => {
        it('should update the task list', () => {
            render(<CardList />);
            fireEvent.change(screen.getByTestId('task_input'), {
                target: { value: 'Task Name' },
            });

            fireEvent.click(screen.getByTestId('add-task_btn'));

            expect(screen.getByText('Task Name')).toBeTruthy();
        });
    });

    
});
