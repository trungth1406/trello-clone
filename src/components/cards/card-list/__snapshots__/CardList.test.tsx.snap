// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When CardList mounted Should render a default card 1`] = `
<ul
  className="card-list"
>
  <li
    className="card-container"
    data-testid="card"
  >
    <h3
      className="card-header"
      data-testid="card_header"
      onClick={[Function]}
    >
      Default Card
    </h3>
    <ul
      className="card-item-list"
      data-testid="task_list"
    />
    <section
      className="card-item-form"
    >
      <input
        data-testid="task_input"
        onChange={[Function]}
        onKeyUp={[Function]}
        placeholder="Add Task"
        type="text"
      />
      <button
        className="add-card-item-button"
        data-testid="add-task_btn"
        onClick={[Function]}
      />
    </section>
    <section
      className="delete-card-button"
      onClick={[Function]}
    >
      <button
        data-testid="delete_btn"
        id="delete_btn"
      >
        Delete
      </button>
    </section>
  </li>
  <button
    className="add-card-button"
    data-testid="add-card_btn"
    id="add-card_btn"
    onClick={[Function]}
  >
    <i
      className="fa fa-plus"
    />
  </button>
</ul>
`;
